File: CPP/005_Program/main.cpp
#include <iostream>
#include <fstream>
#include <string>
using namespace std;

/**
 * File Operations - File I/O Demonstration
 * Shows file reading, writing, and manipulation
 */

void writeToFile(const string& filename, const string& content) {
    ofstream file(filename);

    if (file.is_open()) {
        file << content;
        file.close();
        cout << "Successfully wrote to " << filename << endl;
    } else {
        cout << "Error: Unable to open file for writing!" << endl;
    }
}

void readFromFile(const string& filename) {
    ifstream file(filename);
    string line;

    if (file.is_open()) {
        cout << "\nContent of " << filename << ":" << endl;
        cout << "================================" << endl;

        while (getline(file, line)) {
            cout << line << endl;
        }

        cout << "================================" << endl;
        file.close();
    } else {
        cout << "Error: Unable to open file for reading!" << endl;
    }
}

void appendToFile(const string& filename, const string& content) {
    ofstream file(filename, ios::app);

    if (file.is_open()) {
        file << content;
        file.close();
        cout << "Successfully appended to " << filename << endl;
    } else {
        cout << "Error: Unable to open file for appending!" << endl;
    }
}

int main() {
    cout << "=== C++ File Operations ===" << endl;

    string filename = "sample.txt";

    // Write to file
    cout << "\n1. Writing to file..." << endl;
    string content = "Hello, this is a C++ file I/O example.\n";
    content += "This file demonstrates reading and writing operations.\n";
    content += "C++ makes file handling easy with fstream library.\n";

    writeToFile(filename, content);

    // Read from file
    cout << "\n2. Reading from file..." << endl;
    readFromFile(filename);

    // Append to file
    cout << "\n3. Appending to file..." << endl;
    string appendContent = "\nThis line was appended to the file.\n";
    appendContent += "Appending allows adding content without overwriting.\n";

    appendToFile(filename, appendContent);

    // Read again to see appended content
    cout << "\n4. Reading file after appending..." << endl;
    readFromFile(filename);

    // User input to file
    cout << "\n5. Adding user input to file..." << endl;
    cout << "Enter text to add to file (press Enter to finish): ";

    string userInput;
    getline(cin, userInput);

    if (!userInput.empty()) {
        appendToFile(filename, "\nUser input: " + userInput + "\n");
        readFromFile(filename);
    }

    return 0;
}

File: Ruby/005_Program/main.rb
#!/usr/bin/env ruby

# File Operations - File I/O Demonstration
# Shows file reading, writing, and manipulation

def write_to_file(filename, content)
  File.open(filename, 'w') do |file|
    file.write(content)
  end
  puts "Successfully wrote to #{filename}"
rescue => e
  puts "Error writing to file: #{e.message}"
end

def read_from_file(filename)
  content = File.read(filename)
  puts "\nContent of #{filename}:"
  puts "=" * 40
  puts content
  puts "=" * 40
rescue => e
  puts "Error reading file: #{e.message}"
end

def append_to_file(filename, content)
  File.open(filename, 'a') do |file|
    file.write(content)
  end
  puts "Successfully appended to #{filename}"
rescue => e
  puts "Error appending to file: #{e.message}"
end

def read_lines(filename)
  lines = File.readlines(filename)
  puts "\nFile has #{lines.length} lines:"
  lines.each_with_index do |line, index|
    puts "Line #{index + 1}: #{line.chomp}"
  end
rescue => e
  puts "Error reading file: #{e.message}"
end

def file_info(filename)
  if File.exist?(filename)
    puts "\nFile Information:"
    puts "  Name: #{filename}"
    puts "  Size: #{File.size(filename)} bytes"
    puts "  Readable: #{File.readable?(filename)}"
    puts "  Writable: #{File.writable?(filename)}"
    puts "  Directory: #{File.directory?(filename)}"
    puts "  File: #{File.file?(filename)}"

    stat = File.stat(filename)
    puts "  Created: #{stat.ctime}"
    puts "  Modified: #{stat.mtime}"
  else
    puts "File does not exist: #{filename}"
  end
end

# Main program
puts "=== Ruby File Operations ==="

filename = "sample.txt"

# 1. Write to file
puts "\n1. Writing to file..."
content = <<~TEXT
  Hello, this is a Ruby file I/O example.
  This file demonstrates reading and writing operations.
  Ruby makes file handling elegant and easy.
TEXT

write_to_file(filename, content)

# 2. Read from file
puts "\n2. Reading from file..."
read_from_file(filename)

# 3. Append to file
puts "\n3. Appending to file..."
append_content = <<~TEXT

  This line was appended to the file.
  Appending allows adding content without overwriting.
TEXT

append_to_file(filename, append_content)

# 4. Read file lines
puts "\n4. Reading file by lines..."
read_lines(filename)

# 5. File information
puts "\n5. File information..."
file_info(filename)

# 6. Working with CSV-like data
puts "\n6. Writing structured data..."
users = [
  ["Name", "Age", "City"],
  ["Alice", "25", "New York"],
  ["Bob", "30", "Los Angeles"],
  ["Charlie", "28", "Chicago"]
]

File.open("users.txt", 'w') do |file|
  users.each do |user|
    file.puts user.join(", ")
  end
end

puts "Successfully wrote user data to users.txt"

# Read and display structured data
puts "\nUser Data:"
File.readlines("users.txt").each do |line|
  puts "  #{line.chomp}"
end
